File: backend/python/basics.py
# Haven't used Python for a long time? Here is a quick reminder:

#### Data types
count = 5
weight = 20.5
message = "Hello world"
isReady = True

#### Variable in string
print(f"Hello {userName}")

#### Collections (Data structures)
list = ['Lets', 'Go']
list.append('Hey')
print('Hey' in list) # True

dictionary = {21: "Helga", 22: "Willy"}
dictionary[23] = "John"
print(dictionary[21])

my_set={1, 3} # my_set = set()
my_set.add(7)
print(7 in my_set) # True

#### Null safety
# Warning! No optional chaining
name = print('')
if (name): # name is None
    print(name) # won't be printed

#### Functions
def multiply(number):
    return number * 3
multiply(2)

#### Conditions
# (and, or, not); (==, !=);  (True, False)
if (False):
    print("A")
elif False:
    print("B")
else:
    print("C")

#### Cycles
for user in users:
    print(user)
while True:
    print('hmm')
    # break / continue

#### Exceptions
try:
    raise Exception("No internet")
except Exception as e:
    print(e)
finally:
    print('Done!')

#### Comparison
print(authors == writers) # Structural comparison (true)
print(authors is writers) # Referential comparison (false)

#### Modules (Imports)
import monmod
monmod.printNumbers()

#### Enum
from enum import Enum
class State(Enum):
    IDLE
    RUNNING
print(State.RUNNING)

#### Classes
class Person:
    def __init__(self, name):
        self.user = name

    def greet(self):
        print(f"Hello {self.user}")
person = Person("John")
person.greet()

#### Anonymous class / JSON object
# Warning! It's just a dictionary
status = {"id": 23, "message": "Wrong connection"}
print(status["message"])

#### Lambda (Anonymous function)
# Warning! Multiline lambdas not supported
toUppercase = lambda text: text.upper()
print(toUppercase("hey"))

#### HTTP request/response
import requests
response = requests.get("http://httpbin.org/get")
json = response.json()
print(json["origin"])

#### Run code in parallel
def runInParallel(*fns):
    proc = []
    for fn in fns:
        p = Process(target=fn)
        p.start()
        proc.append(p)
    for p in proc:
        p.join()
runInParallel(functA, functB)

#### Package manager
# Poetry; Files - pyproject.toml, poetry.lock
# Pipenv; Files - Pipfile, Pipfile.lock

#### Naming conventions
# Folder - advanced_query
# File - user_repository.py

File: backend/c#/basics.cs
// Haven't used C# for a long time? Here is a quick reminder:

// Data types
int count = 5;
double weight = 3.5;
string message = "Hello world";
bool isReady = true;

// Variable in string
var userName = "John";
Console.WriteLine($"Hello {userName}");

// Collections
var list = new List<string>();
list.Add("Hey");
bool containsHey = list.Contains("Hey");

var map = new Dictionary<int, string>();
map[23] = "John";
string value = map[23];

var set = new HashSet<int>();
set.Add(2);
bool hasTwo = set.Contains(2);

// Null safety
string address = bob?.Department?.Head?.Address;

string name = null;
if (name != null)
{
    Console.WriteLine(name); // Won't be printed
}

// Functions
int Multiply(int number)
{
    return number * 3;
}

int result = Multiply(2);

// Conditions
// (&&, ||, !); (==, !=); (true, false)
if (true)
{
    Console.WriteLine("A");
}
else
{
    Console.WriteLine("B");
}

// Cycles
var users = new List<string> { "User1", "User2" };
foreach (var user in users)
{
    Console.WriteLine("B");
}

int cakesEaten = 0;
while (true)
{
    cakesEaten++;
    break;
}

// Exceptions
try
{
    throw new Exception("No internet");
}
catch (Exception error)
{
    Console.WriteLine(error.Message);
}
finally
{
    Console.WriteLine("Done!");
}

// Comparison
Console.WriteLine(2 == 3);

var obj1 = new { Field1 = "Value1", Field2 = "Value2" };
var obj2 = new { Field1 = "Value1", Field2 = "Value2" };
Console.WriteLine(Equals(obj1, obj2)); // Structural equality check

// Enum usage
enum State
{
    IDLE,
    RUNNING
}

Console.WriteLine(State.IDLE);

// Classes
class Person
{
    public string Name { get; private set; }

    public Person(string name)
    {
        Name = name;
    }

    public void Greet()
    {
        Console.WriteLine($"Hello {Name}");
    }
}

Person person1 = new Person("John");
person1.Greet();

// Anonymous class / JSON object
var status = new
{
    Id = 23,
    Message = "Wrong connection",
    DebugMessage = new Action(() => Console.WriteLine("NUM - 81723"))
};
Console.WriteLine(status.Message);
status.DebugMessage();

// Lambda (Anonymous function)
Func<string, string> toUppercase = text => text.ToUpper();
Console.WriteLine(toUppercase("hey"));

// HTTP request/response
HttpClient client = new HttpClient();
HttpResponseMessage response = await client.GetAsync("http://httpbin.org/get");
string jsonResponse = await response.Content.ReadAsStringAsync();
var json = JsonSerializer.Deserialize<JsonElement>(jsonResponse);
Console.WriteLine(json.GetProperty("origin"));

// Run code in parallel
// Waits for 0.5 seconds, writes "Hello world!"
static async Task Write(string text, int delay)
{
    await Task.Delay(delay);
    Console.WriteLine(text);
}
Task task1 = Write("World!", 500);
Task task2 = Write("Hello", 400);

await Task.WhenAll(task1, task2);

// Package manager - NuGet

// Naming conventions
// Folder - AdvancedQuery
// File - UserRepository.cs
